src/app/modules/ontology-exploration/ontology-search/ontology-search.component.spec.ts

```diff
@@ -1,25 +1,83 @@
-import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { MatAutocompleteSelectedEvent } from '@angular/material/autocomplete';
+import { of } from 'rxjs';
+import { RecursivePartial, Shallow } from 'shallow-render';
 
+import { OntologyNode } from '../../../core/models/ontology-node';
+import { OntologySearchService, SearchResult } from '../../../core/services/ontology-search/ontology-search.service';
 import { OntologySearchComponent } from './ontology-search.component';
+import { OntologySearchModule } from './ontology-search.module';
+
+
+function fromPartial<T>(partial: RecursivePartial<T>): T {
+  return partial as T;
+}
+
 
 describe('OntologySearchComponent', () => {
```

Comment:
No tests for template/dom events i.e. is the expected method called when mat-autocomplete emits, etc.
Missing tests for some class code branches.

Reply:
I think testing Dom using unit testing is overkill. Such tests should be part of functional tests. If we want to test functionality as whole we should  use **Puppeteer** for the same.
We already have preview URL which can help us in writing functional tests using **Puppeteer**.

https://developers.google.com/web/tools/puppeteer

@bherr2 : Thoughts?

Reply:
Agreed that we need to separate out Unit and Functional testing. The "Angular Way" is to use Protractor <https://www.protractortest.org/#/> and here's a tutorial <https://medium.com/@marcozuccaroli/functional-testing-with-protractor-and-angular-cli-getting-started-c3ddeefbc4da>. We can't take on Functional testing this sprint, but we can start looking into it. It may make things easier when we have a clear separation between the two.